### PR TITLE
Ml 385 convert cost functions to num power

### PR DESCRIPTION
**Pull request by apphp**

### CODE REVIEW

File: src/NeuralNet/CostFunctions/Base/Contracts/CostFunction.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * @category    Machine Learning
+ * @package     Rubix/ML
+ * @author      Samuel Akopyan <leumas.a@gmail.com>
+ */
+interface CostFunction extends Stringable
+{
+    /**
+     * Compute the loss score.
+     *
+     * @internal
```
Comment (SkibidiProduction) [inline src/NeuralNet/CostFunctions/Base/Contracts/CostFunction.php]: I think this could be removed so that the user would have the ability to create their own loss function based on this interface. Same below.

Comment (apphp) [reply src/NeuralNet/CostFunctions/Base/Contracts/CostFunction.php]: Done

File: src/NeuralNet/CostFunctions/CrossEntropy/CrossEntropy.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ * @author      Samuel Akopyan <leumas.a@gmail.com>
+ */
+class CrossEntropy implements ClassificationLoss
+{
+    use ValidatesShapes;
+
+    /**
+     * Compute the loss score.
+     *
+     * L(y, ŷ) = -Σ(y * log(ŷ)) / n
```
Comment (andrewdalpino) [inline src/NeuralNet/CostFunctions/CrossEntropy/CrossEntropy.php:35]: Love how you put the formula here!

File: src/NeuralNet/CostFunctions/CrossEntropy/CrossEntropy.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rubix\ML\NeuralNet\CostFunctions\CrossEntropy;
```
Comment (andrewdalpino) [inline src/NeuralNet/CostFunctions/CrossEntropy/CrossEntropy.php:5]: Are you are carving out this namespace to hold the new classes temporarily until we can get the whole Neural Net subsystem converted to NumPower?

File: src/Traits/ValidatesShapes.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+trait ValidatesShapes
+{
+    /**
+     * Validate that output and target have the same shape.
+     *
+     * @param NDArray $output
+     * @param NDArray $target
+     * @throws InvalidArgumentException
+     */
+    protected function validateShapes(NDArray $output, NDArray $target) : void
```
Comment (andrewdalpino) [inline src/Traits/ValidatesShapes.php]: To me, this function name does not describe what the function really does. I would recommend something like "validateShapesAreEqual()" or better yet, this should be a method on the NDArray object. Something to consider implementing in NumPower @SkibidiProduction.

**Review by andrewdalpino — COMMENTED** [review]
Looks great @apphp! I left a couple comments one on function naming and the other just to clarify our temporary namespacing. Other than that, I think it's good to go. I really liked the additional context you provided by the comments ... I'm generally against comments unless they bring outsized value and you hit it just right with the addition of the mathematical formulas.

**Review by andrewdalpino — APPROVED** [review]

File: src/NeuralNet/CostFunctions/CrossEntropy/CrossEntropy.php (hunk 3 above)
Comment (apphp) [reply src/NeuralNet/CostFunctions/CrossEntropy/CrossEntropy.php:5]: Yes, exactly - like it was done by @SkibidiProduction for activation functions.

File: src/Traits/ValidatesShapes.php (hunk 4 above)
Comment (apphp) [reply src/Traits/ValidatesShapes.php]: Ok, agree. Renamed into "assertSameShape"

Comment (apphp) [reply src/Traits/ValidatesShapes.php]: When we add this method to NDArray object, then we can remove this trait
